### PR TITLE
[patch V2] service: avoid to release refcnt twice on n_on_console

### PR DESCRIPTION
**Pull request by pfliu**
manager_status_printf() print message to console according to the refcnt
of n_on_console. Thus a mistaken refcnt causes boring message when a user
tries to interact with emergency shell.
During my test, the unit "dracut pre-pivot and cleanup hook" experiences
service state: start -> exited, unit state: UNIT_ACTIVATING -> UNIT_ACTIVE,
which causes n_on_console-- twice both in unit_notify() and
service_set_state(). This patch fixes this bug by telling whether
old_state equals new_state or not.

---
v1->v2:
Based on keszybz's comment, reuse the code of refcnt in unit_notify()

### CODE REVIEW

File: src/core/unit.h [hunk 1]
```diff
@@ -793,6 +793,8 @@ int unit_prepare_exec(Unit *u);
 
 void unit_warn_leftover_processes(Unit *u);
 
+extern bool service_exited_release_console(Unit *u);
```
Comment (shawnl) [inline src/core/unit.h:796]: The prototype should be in service.h if the code is in service.c. Then unit.c can #include "service.h".

Comment (pfliu) [reply src/core/unit.h:796]: Oh, I had considered about it. But it will expose the detail of Service to Unit, is it ok from the closure of the design? If it is ok, I will push another version.

Thank you very much for the review.

**Comment by keszybz** [comment]
@pfliu, there's no need to open new PRs. Just force-push to the existing branch when you have a new version.

File: src/core/service.c [hunk 2]
```diff
@@ -947,6 +947,24 @@ static void service_search_main_pid(Service *s) {
                 log_unit_warning_errno(UNIT(s), r, "Failed to watch PID "PID_FMT" from: %m", pid);
 }
 
+bool service_exited_release_console(Unit *u)
+{
```
Comment (keszybz) [inline src/core/service.c]: Please move this parenthesis to previous line:
```C
bool service_exited_release_console(Unit *u) {
```

File: src/core/service.c [hunk 3]
```diff
@@ -947,6 +947,24 @@ static void service_search_main_pid(Service *s) {
...
+        Service *s;
+        if (u->type != UNIT_SERVICE)
+            return false;
+
+        s = SERVICE(u);
```
Comment (keszybz) [inline src/core/service.c]: The SERVICE macro already includes the check if the type is UNIT_SERVICE, so you should skip the type check above, and instead check that `s` is not NULL.

File: src/core/service.c [hunk 4]
```diff
@@ -947,6 +947,24 @@ static void service_search_main_pid(Service *s) {
...
+        /* For remain_after_exit services, let's see if we can "release" the
+         * hold on the console, since unit_notify() only does that in case of
+         * change of state */
+        if (s->state == SERVICE_EXITED &&
```
Comment (keszybz) [inline src/core/service.c]: What about skipping the conditional and just saying
```C
    return s->state == SERVICE_EXITED &&
              ...;
```

File: src/core/unit.c [hunk 5]
```diff
@@ -2308,6 +2308,7 @@ static int unit_log_resources(Unit *u) {
 void unit_notify(Unit *u, UnitActiveState os, UnitActiveState ns, bool reload_success) {
         Manager *m;
         bool unexpected;
+        bool special = service_exited_release_console(u);
```
Comment (keszybz) [inline src/core/unit.c]: Please move this assignment after the `assert`s below.

**Comment by pfliu** [comment]
Hi keszybz,
I adopt all your comments, and git push -f V3

Thanks

File: src/core/service.c [hunk 6]
```diff
@@ -947,6 +947,19 @@ static void service_search_main_pid(Service *s) {
...
+        Service *s;
+
+        s = SERVICE(u);
+        if (!s)
+            return false;
+        /* For remain_after_exit services, let's see if we can "release" the
+         * hold on the console, since unit_notify() only does that in case of
+         * change of state */
+        return s->state == SERVICE_EXITED && s->remain_after_exit &&
+            UNIT(s)->manager->n_on_console > 0
```
Comment (keszybz) [inline src/core/service.c:960]: No semicolon... That doesn't even compile!

File: src/core/unit.c [hunk 7]
```diff
@@ -2308,6 +2308,7 @@ static int unit_log_resources(Unit *u) {
 void unit_notify(Unit *u, UnitActiveState os, UnitActiveState ns, bool reload_success) {
         Manager *m;
         bool unexpected;
+        bool special;
```
Comment (keszybz) [inline src/core/unit.c:2311]: Hmm, this name needs to change to reflect the meaning (`release_console_but_remain` or so).

File: src/core/service.c [hunk 8]
```diff
@@ -947,6 +947,19 @@ static void service_search_main_pid(Service *s) {
                 log_unit_warning_errno(UNIT(s), r, "Failed to watch PID "PID_FMT" from: %m", pid);
 }
 
+bool service_exited_release_console(Unit *u){
+        Service *s;
+
+        s = SERVICE(u);
+        if (!s)
+            return false;
+        /* For remain_after_exit services, let's see if we can "release" the
```
Comment (keszybz) [inline src/core/service.c:956]: This comment needs to be updated. This function is now called from `unit_notify`, so "unit_notify() only does that in case of change of state" is not true anymore.

**Comment by keszybz** [comment]
FTR: #7812 was the previous version.

File: src/core/service.c [hunk 9]
```diff
@@ -947,6 +947,19 @@ static void service_search_main_pid(Service *s) {
                 log_unit_warning_errno(UNIT(s), r, "Failed to watch PID "PID_FMT" from: %m", pid);
 }
 
+bool service_exited_release_console(Unit *u){
+        Service *s;
+
+        s = SERVICE(u);
+        if (!s)
+            return false;
```
Comment (poettering) [inline src/core/service.c:955]: please follow coding style, i.e. multiples of 8ch, see CODING_STYLE

File: src/core/service.c [hunk 10]
```diff
@@ -1002,31 +1015,6 @@ static void service_set_state(Service *s, ServiceState state) {
         if (!IN_SET(state, SERVICE_START_POST, SERVICE_RUNNING, SERVICE_RELOAD))
                 service_stop_watchdog(s);
 
-        /* For the inactive states unit_notify() will trim the cgroup,
-         * but for exit we have to do that ourselves... */
-        if (state == SERVICE_EXITED && !MANAGER_IS_RELOADING(UNIT(s)->manager))
-                unit_prune_cgroup(UNIT(s));
```
Comment (poettering) [inline src/core/service.c:1008]: why did you remove this bit?

File: src/core/unit.h (hunk 1 above)
Comment (poettering) [reply src/core/unit.h:796]: Hmm, let's maybe do this properly: let's add a new function pointer to UnitVTable called "needs_console()" or so, which is then implemented by the Service types, but not for the others. Then, let's add a generic function "unit_needs_console()" that calls into the unit-type-specific implementation if the function ptr is set (which it would be for services), and use a generic fallback that just uses exec_context_may_touch_console() as the current code does (which would be used for all other unit types)

With that in place there would then be a single call that tells us whether the unit currently needs a console, and it would be simple, and generic, without losing the type-agnostic handling in unit.c.

As a second thing I'd then add a bool "console_reffed" or so to the Unit object, and set it whenever a unit has reffed n_on_console. That would make the ref counting much more robust, and would simplify the code too.

Does that make sense? @pfliu any chance you can rework this accordingly?

**Comment by poettering** [comment]
So, I prepped #7991 now which implements what I proposed above. I figured it would be easier if I just hacked that up my proposal myself, in prospect that we want to release v237 soon.

Thank you @pfliu for tracking this down and prepping this patch!

Closing in favour of #7991.

**Comment by pfliu** [comment]
@poettering, your implement in #7991 is a better choice than mine, thanks!
